### PR TITLE
QSimSimulator: memoizing circuit translations

### PR DESCRIPTION
**Pull request by verult**
Fixes https://github.com/quantumlib/qsim/issues/340

This implementation uses a queue as the memoization data structure, and circuit match is based on a linear search through the queue using circuit equality checks.

For simplicity, when there is a cache hit, the circuit does not move to the head of the queue. This is fine for the use cases mentioned in https://github.com/quantumlib/qsim/issues/340.

@95-martin-orion

### CODE REVIEW

File: qsimcirq_tests/qsimcirq_test.py [hunk 1]
```diff
@@ -1376,3 +1376,123 @@ def test_cirq_qsim_global_shift():
...
+
+    # Pick qubits.
+    a, b, c, d = [
+        cirq.GridQubit(0, 0),
+        cirq.GridQubit(0, 1),
+        cirq.GridQubit(1, 1),
+        cirq.GridQubit(1, 0),
+    ]
+
+    # compute_amplitudes()
```
Comment (95-martin-orion) [inline qsimcirq_tests/qsimcirq_test.py:1397]: Could you split these (`compute_amplitudes`, `simulate`, etc.) into separate test methods?

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Looks great overall! Do you have an example of how much this improves performance for many repetitions of (relatively) small circuits?

**Comment by verult** [comment]
Basing off of the cirq quantum phase estimation example, I benchmarked with the code below:

```python
import cirq
import qsimcirq
import timeit
import numpy as np


def run():
    repetitions = 1000
    qnum = 8
    phi = 0.2
    gate = cirq.MatrixGate(matrix=np.array([[np.exp(2 * np.pi * 1.0j * phi), 0], [0, 1]]))

    ancilla = cirq.LineQubit(-1)
    qubits = cirq.LineQubit.range(qnum)

    oracle_raised_to_power = [
        gate.on(ancilla).controlled_by(qubits[i]) ** (2 ** i) for i in range(qnum)
    ]
    circuit = cirq.Circuit(
        cirq.H.on_each(*qubits),
        oracle_raised_to_power,
        cirq.qft(*qubits, without_reverse=True) ** -1,
    )

    simulator = qsimcirq.QSimSimulator(circuit_memoization_size=1)
    #simulator = qsimcirq.QSimSimulator()
    for _ in range(repetitions):
        simulator.simulate(circuit)


print("Execution time: ", timeit.timeit(run, number=1))
```

On my machine: without memoization: ~3.7s, with memoization: ~0.6s.

File: qsimcirq_tests/qsimcirq_test.py (hunk 1 above)
Comment (verult) [reply qsimcirq_tests/qsimcirq_test.py]: done :)

**Review by 95-martin-orion — APPROVED** [review]

**Comment by verult** [comment]
addressed python format failure

**Pull request merged**